In [1]:
import pandas as pd

#import all files
df1 = pd.read_csv("North_America_Airbus.csv") 
df2 = pd.read_csv("North_America_Boeing_CFM.csv")
df3 = pd.read_csv("North_America_Boeing_Rest.csv")
df4 = pd.read_csv("EasternEU_Airbus.csv")
df5 = pd.read_csv("EasternEU_Boeing.csv")
df6 = pd.read_csv("West_EU_Airbus.csv")
df7 = pd.read_csv("West_EU_Boeing.csv")
df8 = pd.read_csv("Latin_Airbus.csv")
df9 = pd.read_csv("Latin_Boeing.csv")
iso = pd.read_csv("paises.csv")

#combine all dataframes into one
df_aw = pd.concat([df1, df2, df3, df4, df5, df6, df7, df8, df9], ignore_index=True)
df_iso = iso



In [2]:
df_aw.head()

Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Aircraft Variant Detail Engine Manufacturer Engine Type  \
0                     212                 CFM       CFM56   
1                     214                 CFM       CFM56   
2                     112                 CFM       CFM56   
3                     214                 CFM       CFM56   
4                     214                 CFM       CFM56   

  Engine Model Variant Engine Variant Detail APU Full Name  \
0                   5A                   5A3             -   
1                   5B                 5B4/P             -   
2                   5B                 5B6/P             -   
3                   5B                 5B4/P             -   
4                   5B                 5B4/P             -   

  Aircraft Registration Number  ...                   Operator Phone  \
0                       N364NW  ...                     404-715-2600   
1                       C-FGJI  ...  888-247-2262 (Customer Service)   
2                       N302NV  ...                     702-851-7300   
3                       N297NV  ...                     702-851-7300   
4                       N283NV  ...                     702-851-7300   

                       Operator Email               Registered Owner/Manager  \
0                                   -                        DELTA AIR LINES   
1                                   -                AERGEN AVIATION FINANCE   
2  customerrelations@allegiantair.com  SUNRISE ASSET MANAGEMENT LLC [NV-USA]   
3  customerrelations@allegiantair.com                 BANK OF UTAH [TRUSTEE]   
4  customerrelations@allegiantair.com  SUNRISE ASSET MANAGEMENT LLC [NV-USA]   

  In Service On Order Stored  Hours Cycles Age (Since Delivery)  \
0        1.0      NaN    NaN  75833  35997                 26.8   
1        1.0      NaN    NaN  70865  27058                 23.6   
2        1.0      NaN    NaN  51115  31015                 20.8   
3        1.0      NaN    NaN  55338  18967                 19.6   
4        1.0      NaN    NaN  53381  21646                 19.2   

  Number of Engines for Operators  
0                               2  
1                               2  
2                               2  
3                               2  
4                               2  

[5 rows x 27 columns]

In [3]:
df_iso.head()

nombre         name          nom  iso2  iso3  phone_code
0  Afganistán  Afghanistan  Afghanistan    AF   AFG          93
1     Albania      Albania      Albanie    AL   ALB         355
2    Alemania      Germany    Allemagne    DE   DEU          49
3     Andorra      Andorra      Andorra    AD   AND         376
4      Angola       Angola       Angola    AO   AGO         244

In [4]:
#Clean column names & database entries: PAISES.CSV
df_iso.columns = df_iso.columns.str.strip()

df_iso['name'] = df_iso['name'].astype('string').str.strip().str.lower()

df_iso.head()


nombre         name          nom iso2 iso3 phone_code
0  Afganistán  afghanistan  Afghanistan   AF  AFG         93
1     Albania      albania      Albanie   AL  ALB        355
2    Alemania      germany    Allemagne   DE  DEU         49
3     Andorra      andorra      Andorra   AD  AND        376
4      Angola       angola       Angola   AO  AGO        244

In [5]:
#Review column required within the exported files
df_iso = df_iso[['name','iso3']]
df_iso.head()

name iso3
0  afghanistan  AFG
1      albania  ALB
2      germany  DEU
3      andorra  AND
4       angola  AGO

In [6]:
#Review column required within the exported files
df_aw = df_aw[['Aircraft Manufacturer', 'Aircraft Type', 'Aircraft Model Variant', 'Engine Manufacturer', 'Engine Type','Engine Model Variant',
       'Aircraft Registration Number','Aircraft Status', 'Aircraft Delivery Date','Operator', 'Operator Country', 'Hours','Cycles', 'Age (Since Delivery)', 'Number of Engines for Operators']]

df_aw.head()

Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Engine Manufacturer Engine Type Engine Model Variant  \
0                 CFM       CFM56                   5A   
1                 CFM       CFM56                   5B   
2                 CFM       CFM56                   5B   
3                 CFM       CFM56                   5B   
4                 CFM       CFM56                   5B   

  Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                       N364NW      In Service             03/04/1999   
1                       C-FGJI      In Service             06/06/2002   
2                       N302NV      In Service             02/07/2005   
3                       N297NV      In Service             05/30/2006   
4                       N283NV      In Service             10/30/2006   

                Operator Operator Country  Hours  Cycles  \
0  Delta Air Lines, Inc.    United States  75833   35997   
1             Air Canada           Canada  70865   27058   
2          Allegiant Air    United States  51115   31015   
3          Allegiant Air    United States  55338   18967   
4          Allegiant Air    United States  53381   21646   

   Age (Since Delivery)  Number of Engines for Operators  
0                  26.8                                2  
1                  23.6                                2  
2                  20.8                                2  
3                  19.6                                2  
4                  19.2                                2

In [7]:
#Extract AC type from Aircraft Type column, where aircraft manufacturer is Airbus left 4 digits in aircraft type and boeing left 3 digits
df_aw['AC Type'] = df_aw.apply(lambda row: row['Aircraft Type'][:4] if 'Airbus' in row['Aircraft Manufacturer'] else row['Aircraft Type'][:3], axis=1)

df_aw.head()


Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Engine Manufacturer Engine Type Engine Model Variant  \
0                 CFM       CFM56                   5A   
1                 CFM       CFM56                   5B   
2                 CFM       CFM56                   5B   
3                 CFM       CFM56                   5B   
4                 CFM       CFM56                   5B   

  Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                       N364NW      In Service             03/04/1999   
1                       C-FGJI      In Service             06/06/2002   
2                       N302NV      In Service             02/07/2005   
3                       N297NV      In Service             05/30/2006   
4                       N283NV      In Service             10/30/2006   

                Operator Operator Country  Hours  Cycles  \
0  Delta Air Lines, Inc.    United States  75833   35997   
1             Air Canada           Canada  70865   27058   
2          Allegiant Air    United States  51115   31015   
3          Allegiant Air    United States  55338   18967   
4          Allegiant Air    United States  53381   21646   

   Age (Since Delivery)  Number of Engines for Operators AC Type  
0                  26.8                                2    A320  
1                  23.6                                2    A320  
2                  20.8                                2    A319  
3                  19.6                                2    A320  
4                  19.2                                2    A320

In [8]:
#Review unique values for AC Type
df_aw['AC Type'].unique()

array(['A320', 'A319', 'A321', 'A330', 'A300', 'A220', 'A350', 'A380',
       '737', '727', '707', '767', '757', '747', '777', 'MD-', '717',
       '787', 'A340', 'A310'], dtype=object)

In [9]:
#Exclude values "MD-" from column AC type
df_aw = df_aw[~df_aw['AC Type'].str.startswith('MD-', na=False)]
df_aw['AC Type'].unique()

array(['A320', 'A319', 'A321', 'A330', 'A300', 'A220', 'A350', 'A380',
       '737', '727', '707', '767', '757', '747', '777', '717', '787',
       'A340', 'A310'], dtype=object)

In [10]:
#Reorder columns
df_aw = df_aw[['Aircraft Manufacturer', 'AC Type', 'Aircraft Model Variant', 'Engine Manufacturer', 'Engine Type','Engine Model Variant',
       'Aircraft Registration Number','Aircraft Status', 'Aircraft Delivery Date','Operator', 'Operator Country', 'Hours','Cycles', 'Age (Since Delivery)', 'Number of Engines for Operators']]
df_aw

Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
19633                Boeing     737                  Max 8   
19634                Boeing     737                 Max 10   
19635                Boeing     737                 Max 10   
19636                Boeing     737                 Max 10   
19637                Boeing     737                 Max 10   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
19633                 CFM        LEAP                   1B   
19634                 CFM        LEAP                   1B   
19635                 CFM        LEAP                   1B   
19636                 CFM        LEAP                   1B   
19637                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             03/04/1999   
1                           C-FGJI      In Service             06/06/2002   
2                           N302NV      In Service             02/07/2005   
3                           N297NV      In Service             05/30/2006   
4                           N283NV      In Service             10/30/2006   
...                            ...             ...                    ...   
19633                       PS-...        On Order             10/15/2033   
19634                       PS-...        On Order             11/15/2033   
19635                       PS-...        On Order             01/15/2034   
19636                       PS-...        On Order             02/15/2034   
19637                       PS-...        On Order             03/15/2034   

                     Operator Operator Country  Hours  Cycles  \
0       Delta Air Lines, Inc.    United States  75833   35997   
1                  Air Canada           Canada  70865   27058   
2               Allegiant Air    United States  51115   31015   
3               Allegiant Air    United States  55338   18967   
4               Allegiant Air    United States  53381   21646   
...                       ...              ...    ...     ...   
19633  GOL Linhas Aéreas S.A.           Brazil      0       0   
19634  GOL Linhas Aéreas S.A.           Brazil      0       0   
19635  GOL Linhas Aéreas S.A.           Brazil      0       0   
19636  GOL Linhas Aéreas S.A.           Brazil      0       0   
19637  GOL Linhas Aéreas S.A.           Brazil      0       0   

       Age (Since Delivery)  Number of Engines for Operators  
0                      26.8                                2  
1                      23.6                                2  
2                      20.8                                2  
3                      19.6                                2  
4                      19.2                                2  
...                     ...                              ...  
19633                   0.0                                2  
19634                   0.0                                2  
19635                   0.0                                2  
19636                   0.0                                2  
19637                   0.0                                2  

[19627 rows x 15 columns]

In [11]:
#Clean database entries for 'Operator Country':
df_aw['Operator Country'] = df_aw['Operator Country'].astype('string').str.strip().str.lower()

df_aw.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             03/04/1999  Delta Air Lines, Inc.   
1      In Service             06/06/2002             Air Canada   
2      In Service             02/07/2005          Allegiant Air   
3      In Service             05/30/2006          Allegiant Air   
4      In Service             10/30/2006          Allegiant Air   

  Operator Country  Hours  Cycles  Age (Since Delivery)  \
0    united states  75833   35997                  26.8   
1           canada  70865   27058                  23.6   
2    united states  51115   31015                  20.8   
3    united states  55338   18967                  19.6   
4    united states  53381   21646                  19.2   

   Number of Engines for Operators  
0                                2  
1                                2  
2                                2  
3                                2  
4                                2

In [12]:
#Merge df_aw with iso to get the ISO3 code for each country in 'Operator Country' column
df_aw = df_aw.merge(df_iso, how='left', left_on='Operator Country', right_on='name')

df_aw.head()


Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             03/04/1999  Delta Air Lines, Inc.   
1      In Service             06/06/2002             Air Canada   
2      In Service             02/07/2005          Allegiant Air   
3      In Service             05/30/2006          Allegiant Air   
4      In Service             10/30/2006          Allegiant Air   

  Operator Country  Hours  Cycles  Age (Since Delivery)  \
0    united states  75833   35997                  26.8   
1           canada  70865   27058                  23.6   
2    united states  51115   31015                  20.8   
3    united states  55338   18967                  19.6   
4    united states  53381   21646                  19.2   

   Number of Engines for Operators           name iso3  
0                                2  united states  USA  
1                                2         canada  CAN  
2                                2  united states  USA  
3                                2  united states  USA  
4                                2  united states  USA

In [13]:
#Reorder columns
df_aw = df_aw = df_aw[['Aircraft Manufacturer', 'AC Type', 'Aircraft Model Variant', 'Engine Manufacturer', 'Engine Type','Engine Model Variant',
       'Aircraft Registration Number','Aircraft Status', 'Aircraft Delivery Date','Operator', 'Operator Country', 'iso3','Hours','Cycles', 'Age (Since Delivery)', 'Number of Engines for Operators']]

df_aw

Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
19622                Boeing     737                  Max 8   
19623                Boeing     737                 Max 10   
19624                Boeing     737                 Max 10   
19625                Boeing     737                 Max 10   
19626                Boeing     737                 Max 10   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
19622                 CFM        LEAP                   1B   
19623                 CFM        LEAP                   1B   
19624                 CFM        LEAP                   1B   
19625                 CFM        LEAP                   1B   
19626                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             03/04/1999   
1                           C-FGJI      In Service             06/06/2002   
2                           N302NV      In Service             02/07/2005   
3                           N297NV      In Service             05/30/2006   
4                           N283NV      In Service             10/30/2006   
...                            ...             ...                    ...   
19622                       PS-...        On Order             10/15/2033   
19623                       PS-...        On Order             11/15/2033   
19624                       PS-...        On Order             01/15/2034   
19625                       PS-...        On Order             02/15/2034   
19626                       PS-...        On Order             03/15/2034   

                     Operator Operator Country iso3  Hours  Cycles  \
0       Delta Air Lines, Inc.    united states  USA  75833   35997   
1                  Air Canada           canada  CAN  70865   27058   
2               Allegiant Air    united states  USA  51115   31015   
3               Allegiant Air    united states  USA  55338   18967   
4               Allegiant Air    united states  USA  53381   21646   
...                       ...              ...  ...    ...     ...   
19622  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19623  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19624  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19625  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19626  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   

       Age (Since Delivery)  Number of Engines for Operators  
0                      26.8                                2  
1                      23.6                                2  
2                      20.8                                2  
3                      19.6                                2  
4                      19.2                                2  
...                     ...                              ...  
19622                   0.0                                2  
19623                   0.0                                2  
19624                   0.0                                2  
19625                   0.0                                2  
19626                   0.0                              

In [14]:
#Convert 'Aircraft Delivery Date' to datetime format and save the year
df_aw['Aircraft Delivery Date'] = pd.to_datetime(df_aw['Aircraft Delivery Date'], errors='coerce', dayfirst=True)
df_aw['Delivery Year'] = df_aw['Aircraft Delivery Date'].dt.year.astype('Int64')

df_aw.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             1999-04-03  Delta Air Lines, Inc.   
1      In Service             2002-06-06             Air Canada   
2      In Service             2005-07-02          Allegiant Air   
3      In Service                    NaT          Allegiant Air   
4      In Service                    NaT          Allegiant Air   

  Operator Country iso3  Hours  Cycles  Age (Since Delivery)  \
0    united states  USA  75833   35997                  26.8   
1           canada  CAN  70865   27058                  23.6   
2    united states  USA  51115   31015                  20.8   
3    united states  USA  55338   18967                  19.6   
4    united states  USA  53381   21646                  19.2   

   Number of Engines for Operators  Delivery Year  
0                                2           1999  
1                                2           2002  
2                                2           2005  
3                                2           <NA>  
4                                2           <NA>

In [15]:
#Filling missing values in 'Delivery Year' with actual year minus age(since delivery)

current_year = pd.Timestamp.now().year

# 1) Asegura tipos numéricos
df_aw['Age (Since Delivery)'] = pd.to_numeric(df_aw['Age (Since Delivery)'], errors='coerce')
df_aw['Delivery Year'] = pd.to_numeric(df_aw['Delivery Year'], errors='coerce')

# 2) Calcula año candidato solo donde falte (puede dar decimales si la edad no es entera)
deliv_year_calc = current_year - df_aw['Age (Since Delivery)']

# 3) Limpia el resultado: rango razonable y redondea a año
deliv_year_calc = deliv_year_calc.where(deliv_year_calc.between(2000, current_year), other=pd.NA)  # descarta valores raros
deliv_year_calc = deliv_year_calc.round()   # o .floor() si prefieres truncar
deliv_year_calc = deliv_year_calc.astype('Int64')  # ya es seguro convertir

# 4) Redondea y convierte también la columna existente (por si vino como float)
deliv_year = df_aw['Delivery Year'].round().astype('Int64')

# 5) Rellena faltantes con el candidato
df_aw['Delivery Year'] = deliv_year.fillna(deliv_year_calc)

df_aw.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             1999-04-03  Delta Air Lines, Inc.   
1      In Service             2002-06-06             Air Canada   
2      In Service             2005-07-02          Allegiant Air   
3      In Service                    NaT          Allegiant Air   
4      In Service                    NaT          Allegiant Air   

  Operator Country iso3  Hours  Cycles  Age (Since Delivery)  \
0    united states  USA  75833   35997                  26.8   
1           canada  CAN  70865   27058                  23.6   
2    united states  USA  51115   31015                  20.8   
3    united states  USA  55338   18967                  19.6   
4    united states  USA  53381   21646                  19.2   

   Number of Engines for Operators  Delivery Year  
0                                2           1999  
1                                2           2002  
2                                2           2005  
3                                2           2005  
4                                2           2006

In [16]:
df_aw['Engine Manufacturer'].unique()

#Replace 'Rolls-Royce Deutshland' with 'Rolls-Royce' in 'Engine Manufacturer' column
df_aw['Engine Manufacturer'] = df_aw['Engine Manufacturer'].replace('Rolls-Royce Deutschland', 'Rolls-Royce')

In [17]:
#check unique values again
df_aw['Engine Manufacturer'].unique()

array(['CFM', 'Pratt & Whitney', 'Rolls-Royce', 'GE'], dtype=object)

In [18]:
#Merge columns Aircraft Manufacturer & Engine Manufaturer in a new column 'ManufacturerAC_Engine'

df_aw['ManufacturerAC_Engine'] = df_aw['Aircraft Manufacturer'] + ' - ' + df_aw['Engine Manufacturer']

df_aw

Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
19622                Boeing     737                  Max 8   
19623                Boeing     737                 Max 10   
19624                Boeing     737                 Max 10   
19625                Boeing     737                 Max 10   
19626                Boeing     737                 Max 10   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
19622                 CFM        LEAP                   1B   
19623                 CFM        LEAP                   1B   
19624                 CFM        LEAP                   1B   
19625                 CFM        LEAP                   1B   
19626                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             1999-04-03   
1                           C-FGJI      In Service             2002-06-06   
2                           N302NV      In Service             2005-07-02   
3                           N297NV      In Service                    NaT   
4                           N283NV      In Service                    NaT   
...                            ...             ...                    ...   
19622                       PS-...        On Order                    NaT   
19623                       PS-...        On Order                    NaT   
19624                       PS-...        On Order                    NaT   
19625                       PS-...        On Order                    NaT   
19626                       PS-...        On Order                    NaT   

                     Operator Operator Country iso3  Hours  Cycles  \
0       Delta Air Lines, Inc.    united states  USA  75833   35997   
1                  Air Canada           canada  CAN  70865   27058   
2               Allegiant Air    united states  USA  51115   31015   
3               Allegiant Air    united states  USA  55338   18967   
4               Allegiant Air    united states  USA  53381   21646   
...                       ...              ...  ...    ...     ...   
19622  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19623  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19624  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19625  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   
19626  GOL Linhas Aéreas S.A.           brazil  BRA      0       0   

       Age (Since Delivery)  Number of Engines for Operators  Delivery Year  \
0                      26.8                                2           1999   
1                      23.6                                2           2002   
2                      20.8                                2           2005   
3                      19.6                                2           2005   
4                      19.2                                2           2006   
...                     ...                              ...            ...   
19622                   0.0                                2           2025   
19623                   0.0                                2           2025   
19624                   0.0            

In [19]:
#filter the aircraft status to In Service
df_aw_service = df_aw[df_aw['Aircraft Status'] == 'In Service']

df_aw_service.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             1999-04-03  Delta Air Lines, Inc.   
1      In Service             2002-06-06             Air Canada   
2      In Service             2005-07-02          Allegiant Air   
3      In Service                    NaT          Allegiant Air   
4      In Service                    NaT          Allegiant Air   

  Operator Country iso3  Hours  Cycles  Age (Since Delivery)  \
0    united states  USA  75833   35997                  26.8   
1           canada  CAN  70865   27058                  23.6   
2    united states  USA  51115   31015                  20.8   
3    united states  USA  55338   18967                  19.6   
4    united states  USA  53381   21646                  19.2   

   Number of Engines for Operators  Delivery Year ManufacturerAC_Engine  
0                                2           1999          Airbus - CFM  
1                                2           2002          Airbus - CFM  
2                                2           2005          Airbus - CFM  
3                                2           2005          Airbus - CFM  
4                                2           2006          Airbus - CFM

In [21]:
import plotly.express as px
import plotly.graph_objects as go

In [22]:
#1. Chart of Aircraft Manufacturer by Aircraft Delivery Year

# Agrupar datos por año y fabricante
evol_manufacturer = (
    df_aw_service.groupby(['Delivery Year', 'Aircraft Manufacturer'])
         .size()
         .reset_index(name='Number of Aircraft')
         .sort_values('Delivery Year')
)

# Crear el gráfico
fig = px.line(
    evol_manufacturer,
    x='Delivery Year',
    y='Number of Aircraft',
    color='Aircraft Manufacturer',
    markers=True,
    title='Evolución del nº aeronaves en servicio por aircraft manufacturer',
    labels={'Delivery Year': 'Año de Entrega', 'Number of Aircraft': 'Nº de Aeronaves'}
)

fig.update_layout(
    xaxis=dict(dtick=1),  # muestra todos los años si no hay demasiados
    legend_title_text='Fabricante'
)
fig.show()

### TENDENCIAS EN AVIONES EN SERVICIO: BOEING VS AIRBUS

**Eventos socio-económicos**
- **COVID**: Caida de las entregas desde finales 2019-2021
- **Tarifas USA**: Subida de los costes y problemas de SC debido a los aranceles impuestos por Trump en 2025.

**Boeing:**
- **Accidentes mortales**: El 737 MAX sufrió dos accidentes fatales en Indonesia (2018) y Etiopía (2019), causando 346 muertes.
- **Pérdida de confianza**: Durante 2019-2020 se observa una caída significativa en las entregas, visible en el gráfico de evolución.
- **Caida de entregas**: La FAA _(Federal Aeroespace Agency)_ impone limite de producción e inspecciones a Boeing.

**Airbus:**
- **Liderazgo en entregas**: Mantiene una tendencia de crecimiento consistente en entregas anuales.
- **Nuevos contratos importantes**: Capta a 3 grandes clientes de la competencia entre 2022-2024.
- **Ramp up de la produccion**: Crea una linea nueva de producción del A320 equivalente al 737 MAX.


In [ ]:
#Group by Engine Manufaturer by Operator Country

#1) Agregación

# a) Recuento DISTINCT por matrícula 
agg = (df_aw_service.dropna(subset=['ManufacturerAC_Engine', 'iso3', 'Operator Country']) 
       .groupby(['iso3','Operator Country', 'ManufacturerAC_Engine'])['Aircraft Registration Number'] 
       .nunique() 
       .reset_index(name='aircraft_count'))

# Filtro top 10 paises:
top_countries = (agg.groupby('iso3')['aircraft_count'].sum() .nlargest(10).index)

agg = agg[agg['iso3'].isin(top_countries)]


#b) Reagrega por motor
agg_engine = (
    df_aw_service.dropna(subset=['iso3','Operator Country','Engine Manufacturer','Aircraft Registration Number'])
         .groupby(['iso3','Operator Country','Engine Manufacturer'])['Aircraft Registration Number']
         .nunique().reset_index(name='aircraft_count')
)

#3) Mapa

fig = px.scatter_geo(
    agg_engine,
    locations='iso3',
    size='aircraft_count',
    color='Engine Manufacturer',
    facet_col='Engine Manufacturer',
    facet_col_wrap=4,         # cuántos por fila
    size_max=24,
    hover_name='Operator Country',
    projection='natural earth',
    title=f'Distribución por país de las top combinaciones de Manufacturer aircraft & Engine'
)


# Limpiar títulos de faceta
fig.for_each_annotation(lambda a: a.update(text=a.text.split('=')[-1]))
fig.update_layout(showlegend=False, margin=dict(l=10, r=10, t=90, b=10))
fig.show()


In [ ]:
import time, random, re, json
from pathlib import Path
from typing import List, Dict, Optional
import requests
from bs4 import BeautifulSoup
import helper

BASE = "https://aviation-safety.net"
CAT  = "ACE"  # Engines
HEADERS = {"User-Agent": "Mozilla/5.0 (master-project; contact: you@example.com)"}
CACHE_DIR = Path("asn_cache"); CACHE_DIR.mkdir(exist_ok=True)



In [ ]:
# --- ejecutar y guardar ---
events_ACE_list = helper.scrape_category_ACE_list_only()

out_fp = Path("Aviation_Safety_Network_ACE_list_only.jsonl")
with out_fp.open("w", encoding="utf-8") as f:
    for e in events_ACE_list:
        f.write(json.dumps(e, ensure_ascii=False) + "\n")

len(events_ACE_list), out_fp.as_posix()

(1142, 'Aviation_Safety_Network_ACE_list_only.jsonl')

In [26]:
import pandas as pd

# 1 Cargar JSONL
df_asn = pd.read_json("Aviation_Safety_Network_ACE_list_only.jsonl", lines=True)

# 2 Convertir fechas sin warnings ni duplicados
df_asn["date"] = pd.to_datetime(
    df_asn["date_text"],
    format="%d %b %Y",  # Ejemplo: 11 Sep 2024
    errors="coerce"
)

# 3 Crear columna de año (solo para las fechas válidas)
df_asn["year"] = df_asn["date"].dt.year.astype("Int64")

# 4 Filtrar rango de años
year1=2000 
year2=2025

df_years_filtered = df_asn[df_asn["year"].between(year1, year2, inclusive="both")].copy()

print(f"Filas totales: {len(df_asn)} | En rango {year1}–{year2}: {len(df_years_filtered)}")


Filas totales: 1142 | En rango 2000–2025: 405


In [41]:
df_years_filtered

date_text               aircraft_text reg_op_text registration  \
736   13 Jan 2000              Shorts 360-300      HB-AAM       HB-AAM   
737   20 Jan 2000             Airbus A320-231      G-VCED       G-VCED   
738   19 Feb 2000  McDonnell Douglas DC-8-63F      N811CK       N811CK   
739   17 Mar 2000             Airbus A330-202      C-GGWA       C-GGWA   
740   24 Mar 2000             Antonov An-12BK    RA-11302     RA-11302   
...           ...                         ...         ...          ...   
1137  13 Mar 2025         Boeing 737-823 (WL)      N885NN       N885NN   
1138  17 Apr 2025         Boeing 737-7BD (WL)      N7724A       N7724A   
1139   6 Jul 2025               Antonov An-2T    RF-00487     RF-00487   
1140   8 Jul 2025               Antonov An-2R    RA-40451     RA-40451   
1141   4 Nov 2025  McDonnell Douglas MD-11(F)      N259UP       N259UP   

                              location_text       date  year reg_merge  
736           Sirte Oil Company, lsf Avisto 2000-01-13  2000    HB-AAM  
737                  Airtours International 2000-01-20  2000    G-VCED  
738                Kitty Hawk International 2000-02-19  2000    N811CK  
739                             Canada 3000 2000-03-17  2000    C-GGWA  
740                     Sky Cabs, lsf Antey 2000-03-24  2000  RA-11302  
...                                     ...        ...   ...       ...  
1137                      American Airlines 2025-03-13  2025    N885NN  
1138                     Southwest Airlines 2025-04-17  2025    N7724A  
1139  DOSAAF Chita Aviation and Sports Club 2025-07-06  2025  RF-00487  
1140                   Zoloto Ynykchana LLC 2025-07-08  2025  RA-40451  
1141            United Parcel Service - UPS 2025-11-04  2025    N259UP  

[405 rows x 8 columns]

In [28]:
# Partimos de df_asn y df_years_filtered ya creados


# 1) Normaliza matrícula (quitar espacios, mayúsculas, mantener guion)
def norm_reg(s: pd.Series) -> pd.Series:
    return (s.astype("string")
              .str.strip()
              .str.upper()
              .str.replace(" ", "", regex=False))

df_asn["reg_merge"] = norm_reg(df_asn["registration"])
df_years_filtered["reg_merge"] = norm_reg(df_years_filtered["registration"])


# 2) Consolidación por matrícula (cuántos accidente, 1º y último)
asn_by_reg = (
    df_years_filtered
      .dropna(subset=["reg_merge", "date"])
      .groupby("reg_merge")
      .agg(
          asn_engine_events=("date", "size"),
          asn_first_event=("date", "min"),
          asn_last_event =("date", "max")
      )
      .reset_index()
)
asn_by_reg



reg_merge  asn_engine_events asn_first_event asn_last_event
0         AGZ                  1      2016-03-09     2016-03-09
1         AKC                  1      2016-08-04     2016-08-04
2         AKG                  1      2010-01-06     2010-01-06
3         AMP                  1      2006-11-15     2006-11-15
4      AP-BHO                  1      2016-12-07     2016-12-07
..        ...                ...             ...            ...
386    ZS-BXF                  1      2003-11-21     2003-11-21
387    ZS-NRJ                  1      2022-01-03     2022-01-03
388    ZS-NRM                  1      2009-09-24     2009-09-24
389    ZS-OEZ                  1      2007-11-07     2007-11-07
390    ZS-PBI                  1      2005-07-11     2005-07-11

[391 rows x 4 columns]

In [29]:
import re
import pandas as pd

# patrones clave
US_N = re.compile(r"^N\d{1,5}[A-Z]{0,2}$")            # N123, N12345, N123AB
HAS_DASH = re.compile(r"-")

def normalize_registration(reg: str) -> str | None:
    """Normaliza matrículas manteniendo los formatos correctos:
       - USA N-nnnnnLL  -> sin guion
       - Canadá C-F/Gxxx -> con guion (inserta si falta)
       - Prefijos de 1 letra (G,F,D,I,O,Y,M,S,T) -> G-ABCD, etc.
       - Prefijos de 2 chars (EC, PH, OO, 9V, 4X, etc.) -> EC-KGJ, 9V-ABC, etc.
    """
    if pd.isna(reg):
        return None
    s = re.sub(r"\s+", "", str(reg).upper())
    s = re.sub(r"-{2,}", "-", s)  # colapsa guiones múltiples

    # si ya lleva guion y no es una rareza, respeta
    if HAS_DASH.search(s):
        # caso Canadá ya correcto C-Fxxx / C-Gxxx
        if re.match(r"^C-[FG][A-Z]{3}$", s):
            return s
        return s

    # USA: N + 1-5 dígitos + 0-2 letras -> sin guion
    if US_N.match(s):
        return s

    # Canadá sin guion: C F/G + 3 letras -> inserta guion tras C
    if re.match(r"^C[FG][A-Z]{3}$", s):
        return f"C-{s[1:]}"

    # Prefijo de 1 letra + 4 letras (GABCD, FXXXX, DXXXX, IXXXX, etc.)
    if len(s) == 5 and s[0] in set("GFDIOYMSTC") and s[1:].isalpha():
        return f"{s[0]}-{s[1:]}"

    # Prefijo de 2 chars (incluye dígitos: 9V, 4X, 5Y...) + 3–5 alfanum
    m = re.match(r"^([A-Z0-9]{2})([A-Z0-9]{3,5})$", s)
    if m and not s.startswith("N"):  # ya cubrimos N arriba
        return f"{m.group(1)}-{m.group(2)}"

    # si no encaja en ninguna regla, devuelve limpio tal cual
    return s

def norm_reg_series(s: pd.Series) -> pd.Series:
    return s.astype("string").apply(normalize_registration)


In [30]:
# 1) Normaliza en tu scrapeo filtrado por años (df_years_filtered)
df_years_filtered["reg_merge"] = norm_reg_series(df_years_filtered["registration"])

asn_by_reg = (
    df_years_filtered
      .dropna(subset=["reg_merge", "date"])
      .groupby("reg_merge", as_index=False)
      .agg(
          asn_engine_events=("date","size"),
          asn_first_event=("date","min"),
          asn_last_event =("date","max"),
      )
)
# 2) Normaliza en tu BBDD df_aw (tu columna se ve como 'Aircraft Registration Number')
df_aw["_reg_merge"] = norm_reg_series(df_aw["Aircraft Registration Number"])

# 3) Merge
df_aw_enriched = (
    df_aw.merge(asn_by_reg, how="left",
                left_on="_reg_merge", right_on="reg_merge")
         .drop(columns=["_reg_merge","reg_merge"])
)
df_aw_enriched["asn_engine_flag"] = df_aw_enriched["asn_engine_events"].fillna(0).gt(0).astype(int)
df_aw_enriched



# 3) Merge
df_aw_enriched = (
    df_aw.merge(asn_by_reg, how="left",
                left_on="_reg_merge", right_on="reg_merge")
         .drop(columns=["_reg_merge","reg_merge"])
)
df_aw_enriched["asn_engine_flag"] = df_aw_enriched["asn_engine_events"].fillna(0).gt(0).astype(int)

df_aw_enriched


Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
19622                Boeing     737                  Max 8   
19623                Boeing     737                 Max 10   
19624                Boeing     737                 Max 10   
19625                Boeing     737                 Max 10   
19626                Boeing     737                 Max 10   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
19622                 CFM        LEAP                   1B   
19623                 CFM        LEAP                   1B   
19624                 CFM        LEAP                   1B   
19625                 CFM        LEAP                   1B   
19626                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             1999-04-03   
1                           C-FGJI      In Service             2002-06-06   
2                           N302NV      In Service             2005-07-02   
3                           N297NV      In Service                    NaT   
4                           N283NV      In Service                    NaT   
...                            ...             ...                    ...   
19622                       PS-...        On Order                    NaT   
19623                       PS-...        On Order                    NaT   
19624                       PS-...        On Order                    NaT   
19625                       PS-...        On Order                    NaT   
19626                       PS-...        On Order                    NaT   

                     Operator  ...  Hours Cycles  Age (Since Delivery)  \
0       Delta Air Lines, Inc.  ...  75833  35997                  26.8   
1                  Air Canada  ...  70865  27058                  23.6   
2               Allegiant Air  ...  51115  31015                  20.8   
3               Allegiant Air  ...  55338  18967                  19.6   
4               Allegiant Air  ...  53381  21646                  19.2   
...                       ...  ...    ...    ...                   ...   
19622  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19623  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19624  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19625  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19626  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   

       Number of Engines for Operators  Delivery Year  ManufacturerAC_Engine  \
0                                    2           1999           Airbus - CFM   
1                                    2           2002           Airbus - CFM   
2                                    2           2005           Airbus - CFM   
3                                    2           2005           Airbus - CFM   
4                                    2           2006           Airbus - CFM   
...                                ...            ...                    ...   
19622                                2           2025           Boeing - CFM   
19623                                2           2025         

In [31]:
# ¿Cuántas matrículas cambiaron con el normalizador (ej. ECKGJ -> EC-KGJ)?
tmp_before = df_aw["Aircraft Registration Number"].str.upper().str.replace(" ", "", regex=False)
tmp_after  = df_aw["_reg_merge"]
print("Matrículas modificadas por la normalización:",
      (tmp_before != tmp_after).sum())

print("Con accidentes en la web ASN:", int(df_aw_enriched["asn_engine_flag"].sum()),
      "| % flota:", f"{100*df_aw_enriched['asn_engine_flag'].mean():.2f}%")


Matrículas modificadas por la normalización: 194
Con accidentes en la web ASN: 11 | % flota: 0.06%


In [32]:
df_aw_enriched

Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
19622                Boeing     737                  Max 8   
19623                Boeing     737                 Max 10   
19624                Boeing     737                 Max 10   
19625                Boeing     737                 Max 10   
19626                Boeing     737                 Max 10   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
19622                 CFM        LEAP                   1B   
19623                 CFM        LEAP                   1B   
19624                 CFM        LEAP                   1B   
19625                 CFM        LEAP                   1B   
19626                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             1999-04-03   
1                           C-FGJI      In Service             2002-06-06   
2                           N302NV      In Service             2005-07-02   
3                           N297NV      In Service                    NaT   
4                           N283NV      In Service                    NaT   
...                            ...             ...                    ...   
19622                       PS-...        On Order                    NaT   
19623                       PS-...        On Order                    NaT   
19624                       PS-...        On Order                    NaT   
19625                       PS-...        On Order                    NaT   
19626                       PS-...        On Order                    NaT   

                     Operator  ...  Hours Cycles  Age (Since Delivery)  \
0       Delta Air Lines, Inc.  ...  75833  35997                  26.8   
1                  Air Canada  ...  70865  27058                  23.6   
2               Allegiant Air  ...  51115  31015                  20.8   
3               Allegiant Air  ...  55338  18967                  19.6   
4               Allegiant Air  ...  53381  21646                  19.2   
...                       ...  ...    ...    ...                   ...   
19622  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19623  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19624  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19625  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   
19626  GOL Linhas Aéreas S.A.  ...      0      0                   0.0   

       Number of Engines for Operators  Delivery Year  ManufacturerAC_Engine  \
0                                    2           1999           Airbus - CFM   
1                                    2           2002           Airbus - CFM   
2                                    2           2005           Airbus - CFM   
3                                    2           2005           Airbus - CFM   
4                                    2           2006           Airbus - CFM   
...                                ...            ...                    ...   
19622                                2           2025           Boeing - CFM   
19623                                2           2025         

In [33]:
# 0) Por si acaso: asegúrate de tener estas variables/funciones creadas
# - norm_reg (normaliza matrículas)
# - asn_by_reg (reg_merge, asn_engine_events, asn_first_event, asn_last_event)

# 1) Sanity checks del merge
print("Filas base:", len(df_aw), "| Filas enriched:", len(df_aw_enriched))
print("Matrículas con accidentes de motor registrados en la web ASN:", int(df_aw_enriched['asn_engine_flag'].sum()))
print("Cobertura con matrícula no nula:", df_aw_enriched['asn_engine_events'].notna().mean().round(3))

# 2) % de tu flota con al menos 1 evento ACE (2000–2025)
pct_flag = 100 * df_aw_enriched['asn_engine_flag'].mean()
print(f"% flota con engine-failure ASN: {pct_flag:.2f}%")

# 3) Top matrículas por nº de eventos (para revisar casos llamativos)
top_regs = (df_aw_enriched
            .loc[df_aw_enriched['asn_engine_events'].notna(), 
                 ['Aircraft Registration Number','asn_engine_events','asn_first_event','asn_last_event']]
            .sort_values('asn_engine_events', ascending=False)
            .head(10))
top_regs

# 4) Agregado por Engine Manufacturer (si lo tienes en df_aw)
if 'Engine Manufacturer' in df_aw_enriched.columns:
    events_by_engine = (df_aw_enriched
                        .groupby('Engine Manufacturer', dropna=False)['asn_engine_events']
                        .sum(min_count=1)
                        .reset_index()
                        .sort_values('asn_engine_events', ascending=False)
                        .head(10))
    events_by_engine


Filas base: 19627 | Filas enriched: 19627
Matrículas con accidentes de motor registrados en la web ASN: 11
Cobertura con matrícula no nula: 0.001
% flota con engine-failure ASN: 0.06%


In [34]:
#Países con mayor proporción de incidentes por aircraft

(
    df_aw_enriched
    .groupby("Operator Country")["asn_engine_flag"]
    .mean()
    .mul(100)
    .round(2)
    .astype(str)
    .add("%")
    .sort_values(ascending=False)
    .head(3)
)



Operator Country
switzerland        0.6%
united kingdom    0.15%
united states      0.1%
Name: asn_engine_flag, dtype: object

In [35]:
#fabricantes con más incidentes relativos

(
    df_aw_enriched
    .groupby("Engine Manufacturer")["asn_engine_flag"]
    .mean()
    .mul(100)
    .round(2)
    .astype(str)
    .add("%")
    .sort_values(ascending=False)
    .head(5)
)


Engine Manufacturer
Pratt & Whitney    0.15%
GE                 0.09%
CFM                0.03%
Rolls-Royce         0.0%
Name: asn_engine_flag, dtype: object